File: packages/rn/ios/DashOtaCrypto.swift
import Foundation
import CryptoKit

/// Trust-critical crypto for iOS, mirroring the shared `openRelease` reference using Apple's
/// CryptoKit (package-based): Ed25519 verification, AES-256-GCM decryption, SHA-256, HMAC, and
/// the SOA1 archive format.
enum DashOtaCrypto {
  static func b64(_ s: String) -> Data? { Data(base64Encoded: s) }

  /// Verify against any embedded key (key ring); true if one validates.
  static func ed25519VerifyAny(publicKeysB64: [String], message: Data, signature: Data) -> Bool {
    for key in publicKeysB64 where !key.isEmpty {
      guard let raw = b64(key) else { continue }
      if let pub = try? Curve25519.Signing.PublicKey(rawRepresentation: raw),
         pub.isValidSignature(signature, for: message) {
        return true
      }
    }
    return false
  }

  /// AES-256-GCM decrypt (throws if the tag fails to authenticate).
  static func aesGcmDecrypt(key: Data, iv: Data, ciphertext: Data, tag: Data) throws -> Data {
    let box = try AES.GCM.SealedBox(nonce: try AES.GCM.Nonce(data: iv), ciphertext: ciphertext, tag: tag)
    return try AES.GCM.open(box, using: SymmetricKey(data: key))
  }

  static func sha256Hex(_ data: Data) -> String { hex(Data(SHA256.hash(data: data))) }

  static func hmacSha256Hex(key: Data, message: Data) -> String {
    let mac = HMAC<SHA256>.authenticationCode(for: message, using: SymmetricKey(data: key))
    return hex(Data(mac))
  }

  static func hex(_ data: Data) -> String { data.map { String(format: "%02x", $0) }.joined() }

  /// Unpack a "SOA1" archive: magic(4) + uint32BE headerLen + header JSON [{path,size}] + blobs.
  static func unpackArchive(_ buf: Data) throws -> [(path: String, data: Data)] {
    guard buf.count >= 8, String(data: buf.subdata(in: 0..<4), encoding: .ascii) == "SOA1" else {
      throw DashOtaError.message("bad archive magic")
    }
    let headerLen = Int(buf[4]) << 24 | Int(buf[5]) << 16 | Int(buf[6]) << 8 | Int(buf[7])
    let headerEnd = 8 + headerLen
    guard buf.count >= headerEnd else { throw DashOtaError.message("truncated archive header") }
    let headerJson = try JSONSerialization.jsonObject(with: buf.subdata(in: 8..<headerEnd))
    guard let entries = headerJson as? [[String: Any]] else { throw DashOtaError.message("bad archive header") }
    var out: [(String, Data)] = []
    var offset = headerEnd
    for entry in entries {
      guard let path = entry["path"] as? String, let size = entry["size"] as? Int else {
        throw DashOtaError.message("bad archive entry")
      }
      guard buf.count >= offset + size else { throw DashOtaError.message("truncated archive blob") }
      out.append((path, buf.subdata(in: offset..<(offset + size))))
      offset += size
    }
    return out
  }
}

/// Simple error type carrying a code/message for the TurboModule promise rejection.
enum DashOtaError: Error {
  case message(String)
  var text: String { if case .message(let m) = self { return m } else { return "dash-ota error" } }
}

File: packages/rn/android/src/main/java/com/dashota/DashOtaCrypto.kt
package com.dashota

import android.util.Base64
import com.google.crypto.tink.subtle.Ed25519Verify
import org.json.JSONArray
import java.security.MessageDigest
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Trust-critical crypto, mirroring the shared `openRelease` reference. Ed25519 verification
 * uses Google **Tink** (package-based) with the raw 32-byte embedded public key; AES-256-GCM,
 * SHA-256 and HMAC use the JDK directly. SOA1 is the package archive format.
 */
object DashOtaCrypto {
  fun b64(s: String): ByteArray = Base64.decode(s, Base64.DEFAULT)

  /** Verify against any embedded key (key ring); true if one validates. */
  fun ed25519VerifyAny(publicKeysB64: List<String>, message: ByteArray, signature: ByteArray): Boolean {
    for (key in publicKeysB64) {
      if (key.isBlank()) continue
      try {
        Ed25519Verify(b64(key)).verify(signature, message)
        return true
      } catch (_: Exception) {
        // try the next key in the ring
      }
    }
    return false
  }

  /** AES-256-GCM decrypt (throws if the tag fails to authenticate). */
  fun aesGcmDecrypt(key: ByteArray, iv: ByteArray, ciphertext: ByteArray, tag: ByteArray): ByteArray {
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, iv))
    val combined = ByteArray(ciphertext.size + tag.size)
    System.arraycopy(ciphertext, 0, combined, 0, ciphertext.size)
    System.arraycopy(tag, 0, combined, ciphertext.size, tag.size)
    return cipher.doFinal(combined)
  }

  fun sha256(bytes: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(bytes)
  fun sha256Hex(bytes: ByteArray): String = toHex(sha256(bytes))

  fun hmacSha256Hex(key: ByteArray, message: ByteArray): String {
    val mac = Mac.getInstance("HmacSHA256")
    mac.init(SecretKeySpec(key, "HmacSHA256"))
    return toHex(mac.doFinal(message))
  }

  fun toHex(bytes: ByteArray): String {
    val sb = StringBuilder(bytes.size * 2)
    for (b in bytes) sb.append(String.format("%02x", b))
    return sb.toString()
  }

  /** Unpack a "SOA1" archive: magic(4) + uint32BE headerLen + header JSON [{path,size}] + blobs. */
  fun unpackArchive(buf: ByteArray): List<Pair<String, ByteArray>> {
    require(buf.size >= 8 && String(buf, 0, 4, Charsets.US_ASCII) == "SOA1") { "bad archive magic" }
    val headerLen = ((buf[4].toInt() and 0xff) shl 24) or
      ((buf[5].toInt() and 0xff) shl 16) or
      ((buf[6].toInt() and 0xff) shl 8) or
      (buf[7].toInt() and 0xff)
    val headerEnd = 8 + headerLen
    require(buf.size >= headerEnd) { "truncated archive header" }
    val header = JSONArray(String(buf, 8, headerLen, Charsets.UTF_8))
    val out = ArrayList<Pair<String, ByteArray>>(header.length())
    var offset = headerEnd
    for (i in 0 until header.length()) {
      val entry = header.getJSONObject(i)
      val size = entry.getInt("size")
      require(buf.size >= offset + size) { "truncated archive blob" }
      out.add(Pair(entry.getString("path"), buf.copyOfRange(offset, offset + size)))
      offset += size
    }
    return out
  }
}
